File: src/main/java/x/samantha/commandsjda/handlers/CommandHandler.java
package x.samantha.commandsjda.handlers;

import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import x.samantha.commandsjda.abs.CommandContext;
import x.samantha.commandsjda.abs.ICommand;
import x.samantha.commandsjda.abs.Module;
import x.samantha.commandsjda.managers.CommandManager;

import java.text.MessageFormat;
import java.util.List;

public class CommandHandler extends ListenerAdapter {

    private final Logger log = LoggerFactory.getLogger(CommandHandler.class);

    Boolean canBeSelf;
    Boolean canBeBot;
    String prefix;
    Boolean enableUnknownCommandMessage;
    List<ICommand> commands;
    List<Module> modules;

    CommandHandler(Boolean canBeSelf, Boolean canBeBot, String prefix, Boolean enableUnknownCommandMessage) {
        this.canBeSelf = canBeSelf;
        this.canBeBot = canBeBot;
        this.prefix = prefix;
        this.enableUnknownCommandMessage = enableUnknownCommandMessage;
    }

    @Override
    public void onGuildMessageReceived(GuildMessageReceivedEvent event) {
        log.info(MessageFormat.format("[{0}] {1}:\t\"{2}\"", event.getGuild().getName(), event.getAuthor().getName(), event.getMessage().getContentRaw()));

        if (!canBeSelf && event.getAuthor() == event.getJDA().getSelfUser()) return;

        if (!canBeBot && event.getAuthor().isBot()) return;

        CommandContext ctx = CommandContext.parse(event, prefix, enableUnknownCommandMessage);

        if (ctx == null) return;

        CommandManager.onCall(ctx);
    }

    public CommandHandler canBeSelfUser(Boolean value) {
        this.canBeSelf = value;
        return this;
    }

    public CommandHandler canBeBot(Boolean value) {
        this.canBeBot = value;
        return this;
    }

    public CommandHandler enableUnknownCommandMessage(Boolean value) {
        this.enableUnknownCommandMessage = value;
        return this;
    }

    public CommandHandler prefix(String value) {
        this.prefix = value;
        return this;
    }

    public CommandHandler commands(List<ICommand> value) {
        this.commands = value;
        return this;
    }

    public CommandHandler modules(List<Module> value) {
        this.modules = value;
        return this;
    }

    public CommandHandler registerModules() {
        this.modules.forEach(CommandManager::registerModule);
        return this;
    }

    public CommandHandler registerCommands() {
        this.commands.forEach(CommandManager::registerCommand);
        return this;
    }
}

File: src/main/kotlin/x/samantha/commandsjda/handlers/CommandHandler.kt
package x.samantha.commandsjda.handlers

import x.samantha.commandsjda.managers.CommandManager
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter
import x.samantha.commandsjda.abs.CommandContext
import x.samantha.commandsjda.abs.ICommand
import x.samantha.commandsjda.abs.Module
import org.slf4j.LoggerFactory

import java.text.MessageFormat
import java.util.Objects

class CommandHandler

internal constructor(private val canBeSelf: Boolean, private val canBeBot: Boolean, private val prefix: String, private val enableUnknownCmdMsg: Boolean) : ListenerAdapter() {

    private val log = LoggerFactory.getLogger(CommandHandler::class.java)

    override fun onMessageReceived(event: MessageReceivedEvent) {
        log.info(MessageFormat.format("[{0}] {1}:\t\"{2}\"",
                if (event!!.guild == null) "PRIVATE" else event.guild.name, event.author.name, event.message.contentRaw))

        if (event.privateChannel != null) return

        if (event.author.isFake) return

        if (!canBeSelf && event.author == event.jda.selfUser) return

        if (!canBeBot && event.author.isBot) return

        val ctx = CommandContext.parse(event, prefix, enableUnknownCmdMsg)
        if (Objects.isNull(ctx)) return

        CommandManager.onCall(ctx!!)
    }

    data class Builder (
        var canBeSelfUser: Boolean = false,
        var canBeBot: Boolean = false,
        var enableUnknownCmdMsg: Boolean = false,
        var prefix: String = "!",
        var commands: MutableList<ICommand> = mutableListOf(),
        var modules: MutableList<Module> = mutableListOf()
    ) {
        /**
         * Add a command to the builder, this will be added to the command registry when Builder#build() is called
         * @param command
         * The command to be added to the Builder
         */
        fun addCommand(command: ICommand): Builder {
            commands.add(command)
            return this
        }

        /**
         * Add a module to the builder, this will be added to the module registry when Builder#build() is called
         * @param module
         * The module to be added to the Builder
         */
        fun addModule(module: Module): Builder {
            modules.add(module)
            return this
        }

        /**
         * Create a new CommandHandler instance with the provided options, commands and modules
         */
        fun build(): CommandHandler {
            this.commands.forEach { CommandManager.registerCommand(it) }
            this.modules.forEach { CommandManager.registerModule(it) }
            return CommandHandler(this.canBeSelfUser, this.canBeBot, this.prefix, this.enableUnknownCmdMsg)
        }
    }
}
